File: dsa-problems/Recursion/solutions/python/REC-008-alternating-vowel-consonant-ladder.py
def get_alternating_permutations(s: str) -> list[str]:
    def is_vowel(c):
        return c in 'aeiou'
    results = []
    used = [False] * len(s)
    def backtrack(current):
        if len(current) == len(s):
            results.append(current)
            return
        for i in range(len(s)):
            if not used[i]:
                if len(current) == 0 or is_vowel(current[-1]) != is_vowel(s[i]):
                    used[i] = True
                    backtrack(current + s[i])
                    used[i] = False
    backtrack("")
    return sorted(results)

def main():
    import sys
    s = sys.stdin.read().strip()
    results = get_alternating_permutations(s)
    if results:
        for perm in results:
            print(perm)
    else:
        print("NONE")

if __name__ == "__main__":
    main()

File: dsa-problems/Recursion/solutions/javascript/REC-008-alternating-vowel-consonant-ladder.js
const readline = require('readline');
const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
let tokens = [];
rl.on('line', (line) => { tokens.push(...line.trim().split(/\s+/)); });
rl.on('close', () => {
    if(tokens.length===0) return;
    let ptr = 0;
    const s = tokens[ptr++];
    
    const sol = new Solution();
    const res = sol.getAlternatingPermutations(s);
    
    if(res.length === 0) {
        console.log("NONE");
    } else {
        res.forEach(p => console.log(p));
    }
});

class Solution {
    getAlternatingPermutations(s) {
        const results = new Set();
        const used = new Array(s.length).fill(false);
        const chars = s.split('');
        
        const isVowel = (c) => "aeiou".includes(c);
        
        const backtrack = (current) => {
            if (current.length === s.length) {
                results.add(current);
                return;
            }
            
            const lastChar = current.length > 0 ? current[current.length - 1] : null;
            const lastIsVowel = lastChar ? isVowel(lastChar) : false;
            
            for (let i = 0; i < s.length; i++) {
                if (!used[i]) {
                    const nextChar = chars[i];
                    const nextIsVowel = isVowel(nextChar);
                    
                    if (current.length === 0 || lastIsVowel !== nextIsVowel) {
                        used[i] = true;
                        backtrack(current + nextChar);
                        used[i] = false;
                    }
                }
            }
        };
        
        backtrack("");
        const sortedRes = Array.from(results).sort();
        return sortedRes;
    }
}
